File: panel/assets/databases.js
// One click, and a Postgres or a Redis exists.
//
// The engine was here the whole time: the Database CRD carries both, with
// backups and a restore rehearsal Coolify does not have. What was missing was
// a way to ask for one that is not kubectl. docs/PLAN.md §7 counted it as the
// highest-return item for that reason.
//
// Nothing here runs on load. app.js calls mountDatabases; the guards at the
// bottom are the ones every view in this directory carries.

function databasesUrl(base) {
  return `${base}/databases`;
}

// The credentials row, and it is the settings page's secret row wearing a
// different hat.
//
// A database password is in exactly the class that decision was about: the
// operator mints it into a Secret the tenant's pods read, git carries the name
// of that Secret and never the value. So it gets the same sentence, for the
// same two reasons — there is no rollback to a value git never held, and a
// Secret deleted by hand is restored by nothing, unlike a manifest that Argo CD
// puts back in about ten seconds.
//
// The endpoint cannot send the password even by accident: its response shape
// has no field for one. That is worth saying on the screen rather than only in
// a comment, because "not shown" and "not there" look identical to a reader
// and only one of them is a promise.
const CREDENTIALS_NOTE =
  "the username and password are generated into a Secret your application reads. " +
  "They are not in git and this page never sees them: there is no rollback to a " +
  "value git never held, and if the Secret is deleted nothing puts it back";

function credentialsView(db) {
  return {
    state: "secret",
    shown: false,
    secretName: (db && db.secretName) || "",
    text: db && db.secretName
      ? `in the Secret ${db.secretName}`
      : "in a Secret this page does not read",
    warning: CREDENTIALS_NOTE,
    inGit: false,
  };
}

// What one database costs, said as a fact rather than as a prediction.
//
// The endpoint explains at length why it cannot forecast the refusal — it may
// not read persistent volume claims or quotas, a namespace may hold several
// apps whose manifests live in other directories, and the quota is applied when
// Argo CD syncs rather than during the request. So the page says the same thing
// the endpoint says: what this takes, what the ceiling is, and that the ceiling
// is shared.
//
// Three screens, because "you have room", "this is the last one" and "this will
// not fit" are three different things to be told and the middle one is the one
// worth seeing before pressing the button.
function claimsView(claims, wanted = 1) {
  const budget = Number((claims && claims.budget) || 0);
  const used = Number((claims && claims.usedByTheseHere) || 0);
  const left = budget - used;
  if (budget === 0) {
    return { state: "unknown", text: "", ok: null, left: null };
  }
  if (wanted > left) {
    return {
      state: "over",
      text: `this would take ${wanted} of the ${left} volume ${left === 1 ? "claim" : "claims"} ` +
        `left here, and the namespace allows ${budget} in total. The commit will be made ` +
        "and the cluster will refuse it when Argo CD syncs",
      ok: false,
      left,
    };
  }
  if (wanted === left) {
    return {
      state: "last",
      text: `this takes the last of the ${budget} volume claims this namespace allows`,
      ok: null,
      left,
    };
  }
  return {
    state: "room",
    text: `this takes ${wanted} of ${budget} volume claims; ${left} are free here`,
    ok: true,
    left,
  };
}

// Backups are two claims, not one, and the page says so before the choice is
// made rather than after the quota refuses.
function claimsWanted(form) {
  return form && form.backups && form.engine !== "redis" ? 2 : 1;
}

// Redis has no backups here, and the reason is not a limitation to apologise
// for: the schedule would run pg_dump against a server that has never heard of
// it. The API refuses the combination, so the form does not offer it.
function backupsOffered(engine) {
  return engine !== "redis";
}

// What a refusal says, in the server's words.
function refusalText(status, body) {
  const detail = body && typeof body.detail === "string" ? body.detail : "";
  if (detail) return detail;
  switch (status) {
    case 403:
      return "removing a database is owner-only, and this account is not an owner";
    case 404:
      return "this app and environment have no repository configured yet";
    default:
      return `the request was refused with ${status} and no detail`;
  }
}

// What a removal says afterwards.
//
// The endpoint sends the sentence, because every clause in it was measured
// against the operator: which objects are owned and go, which volumes stay, and
// that the claims stay spent. Printing it rather than keeping a second copy is
// the rule the catalogue screen set.
//
// A removal that said only "removed" would be the failure this whole response
// exists to prevent: somebody deleting a database to make room, and getting
// none.
function removedText(body) {
  const note = body && typeof body.note === "string" ? body.note : "";
  return note || "the database was withdrawn. Its volumes are not deleted by this.";
}

// mountDatabases renders the view into el and returns a stop function.
function mountDatabases(el, base, options = {}) {
  const { fetcher = globalThis.fetch } = options;
  const doc = el.ownerDocument;
  let stopped = false;
  let claims = null;

  const node = (tag, attrs = {}, ...kids) => {
    const n = doc.createElement(tag);
    for (const [k, v] of Object.entries(attrs)) n.setAttribute(k, v);
    n.append(...kids.flat());
    return n;
  };

  const status = doc.createElement("p");
  status.className = "muted";
  const list = doc.createElement("div");
  list.className = "db-list";
  const form = doc.createElement("form");
  form.className = "db-form";
  const errors = doc.createElement("ul");
  errors.className = "refusal";
  errors.hidden = true;
  el.replaceChildren(status, list, form, errors);

  const showRefusal = (text) => {
    errors.replaceChildren(...(text ? [node("li", {}, text)] : []));
    errors.hidden = !text;
  };

  const field = (name, label, value, placeholder = "") => {
    const input = doc.createElement("input");
    input.type = "text";
    input.name = name;
    input.value = value;
    input.placeholder = placeholder;
    input.setAttribute("aria-label", label);
    input.setAttribute("autocomplete", "off");
    return input;
  };

  const engine = doc.createElement("select");
  engine.setAttribute("aria-label", "Engine");
  for (const value of ["postgres", "redis"]) {
    const option = doc.createElement("option");
    option.value = value;
    option.textContent = value;
    engine.append(option);
  }
  const name = field("name", "Name", "", "main");
  const image = field("image", "Image", "postgres:17.2");
  const storage = field("storage", "Storage", "5Gi");
  const backups = doc.createElement("input");
  backups.type = "checkbox";
  backups.setAttribute("aria-label", "Take backups");
  const backupStorage = field("backupStorage", "Backup volume", "2Gi");
  const cost = doc.createElement("p");
  cost.className = "db-cost";
  const create = doc.createElement("button");
  create.type = "submit";
  create.textContent = "Create";

  const backupsLabel = node("label", {}, backups, "backups, with a restore rehearsal");
  form.replaceChildren(engine, name, image, storage, backupsLabel, backupStorage, create, cost);

  const paintCost = () => {
    // Redis is not offered backups, rather than offered them and refused.
    const offered = backupsOffered(engine.value);
    backupsLabel.hidden = !offered;
    backupStorage.hidden = !offered || !backups.checked;
    const view = claimsView(claims, claimsWanted({ engine: engine.value, backups: backups.checked }));
    cost.textContent = view.text;
    cost.dataset.ok = String(view.ok);
    cost.dataset.state = view.state;
  };
  engine.addEventListener("change", paintCost);
  backups.addEventListener("change", paintCost);

  const draw = (databases) => {
    list.replaceChildren(...databases.map((db) => {
      const creds = credentialsView(db);
      const row = node("div", { class: "db-row", "data-engine": db.engine });
      const remove = doc.createElement("button");
      remove.type = "button";
      remove.className = "link";
      remove.textContent = "Remove";
      remove.addEventListener("click", () => void withdraw(db.name));

      row.append(
        node("strong", {}, db.name),
        node("span", { class: "muted" }, `${db.engine} · ${db.image} · ${db.storage}`),
        node("span", { class: "muted" },
          `${db.claims} volume ${db.claims === 1 ? "claim" : "claims"}`),
        remove,
        node("p", { class: "db-credentials", "data-state": creds.state },
          `${creds.text} — ${creds.warning}`),
      );
      if (db.backups) {
        row.append(node("p", { class: "muted" },
          `backups ${db.backups.schedule}, kept ${db.backups.retainDays} days` +
          (db.backups.rehearse ? ", restored and counted to prove they work" : "")));
      }
      return row;
    }));
  };

  async function load() {
    status.textContent = "loading…";
    status.dataset.ok = "null";
    try {
      const response = await fetcher(databasesUrl(base), { method: "GET" });
      const body = await response.json().catch(() => null);
      if (!response.ok) {
        showRefusal(refusalText(response.status, body));
        status.textContent = "these databases could not be read";
        status.dataset.ok = "false";
        return;
      }
      const databases = (body && body.databases) || [];
      claims = (body && body.claims) || null;
      status.textContent = databases.length
        ? `${databases.length} ${databases.length === 1 ? "database" : "databases"}`
        : "no databases yet";
      status.dataset.ok = "true";
      draw(databases);
      paintCost();
    } catch (err) {
      showRefusal(String((err && err.message) || err));
      status.textContent = "these databases could not be read";
      status.dataset.ok = "false";
    }
  }

  async function withdraw(dbName) {
    if (stopped) return;
    showRefusal("");
    try {
      const response = await fetcher(`${databasesUrl(base)}/${encodeURIComponent(dbName)}`,
        { method: "DELETE" });
      const body = await response.json().catch(() => null);
      if (!response.ok) {
        showRefusal(refusalText(response.status, body));
        return;
      }
      // The server's sentence about what became of the data, kept on screen
      // rather than flashed: it is the answer to a question the reader has not
      // asked yet.
      status.textContent = removedText(body);
      status.dataset.ok = "null";
      await load();
    } catch (err) {
      showRefusal(String((err && err.message) || err));
    }
  }

  form.addEventListener("submit", async (event) => {
    event.preventDefault();
    if (stopped) return;
    create.disabled = true;
    showRefusal("");
    try {
      const wantBackups = backupsOffered(engine.value) && backups.checked;
      const response = await fetcher(databasesUrl(base), {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          name: name.value, engine: engine.value, image: image.value,
          storage: storage.value,
          ...(wantBackups ? { backups: { storage: backupStorage.value, rehearse: true } } : {}),
        }),
      });
      const body = await response.json().catch(() => null);
      if (!response.ok) {
        showRefusal(refusalText(response.status, body));
        return;
      }
      name.value = "";
      await load();
    } catch (err) {
      showRefusal(String((err && err.message) || err));
    } finally {
      create.disabled = false;
    }
  });

  load();

  return function stop() {
    stopped = true;
  };
}

// The page's entry point:
//
//   index.html:  <script src="/databases.js"></script>   before app.js
//   app.js:      window.damgaDatabases.mountDatabases(box, base(), {});
if (typeof window !== "undefined") {
  window.damgaDatabases = {
    mountDatabases, databasesUrl, credentialsView, claimsView, claimsWanted,
    backupsOffered, refusalText, removedText,
  };
}

if (typeof module !== "undefined") {
  module.exports = {
    mountDatabases, databasesUrl, credentialsView, claimsView, claimsWanted,
    backupsOffered, refusalText, removedText, CREDENTIALS_NOTE,
  };
}

File: server/databases.go
/*
Copyright 2026 Orhan Yavuz.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

package server

import (
	"encoding/json"
	"errors"
	"fmt"
	"maps"
	"net/http"
	"regexp"
	"slices"
	"strings"

	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	platformv1alpha1 "github.com/damgahq/damga/api/v1alpha1"
	"github.com/damgahq/damga/authz"
	"github.com/damgahq/damga/evidence"
	"github.com/damgahq/damga/internal/gitwrite"
	"github.com/damgahq/damga/internal/manifest"
	"github.com/damgahq/damga/placement"
)

// The claim budget, and why this endpoint states it rather than enforcing it.
//
// A tenant namespace is fenced at four persistent volume claims
// (internal/manifest/fence.go). A Database takes one for its data, and a second
// for its archives when backups are on — measured in the operator:
// databaseClaim is always rendered, desiredBackupClaim only when Backup is set,
// and the API refuses backups on redis outright. So two backed-up Postgres
// databases use the whole budget, and they share it with every application
// volume in the same namespace.
//
// It cannot predict the refusal, and that is three separate findings rather
// than one:
//
//   - The control plane may not look. Its ClusterRole grants pods, pods/log,
//     pods/exec, deployments, workloads and databases; it carries no rule for
//     persistentvolumeclaims and none for resourcequotas, so live usage is not
//     readable even in principle.
//   - Counting the committed manifests instead would undercount. placement.go
//     says a tenant "may put as many apps in one [namespace] as it likes", and
//     each app's manifests live in its own directory — so the one directory
//     this endpoint reads is not the namespace.
//   - The refusal does not happen during the request anyway. The write path is
//     git: this returns once a commit is pushed, and the quota is applied by
//     the API server when Argo CD syncs. gitwrite says it plainly where it
//     returns a pending record — "this package knows a commit was pushed and
//     nothing more".
//
// So the honest thing is a fact rather than a forecast: what this object will
// take, and what the ceiling is. The refusal, when it comes, arrives where
// every other rejected sync arrives.
const (
	claimBudget      = 4
	claimsForData    = 1
	claimsForBackups = 1
)

// The refusals this endpoint tells apart. One message for all of them would be
// true of each and useful for none: the next move is different every time.
var (
	errDatabaseExists = errors.New("that database already exists here")
	errNoSuchDatabase = errors.New("no database by that name is committed here")
	// errLastManifest: removing this file would leave the directory empty, and
	// gitwrite refuses an empty render because it cannot tell one from a render
	// that failed. Reported as itself rather than as "nothing to write".
	errLastManifest = errors.New(
		"this is the only manifest committed here, and removing it would leave the " +
			"directory empty; unregister the app instead")
)

// databaseFrom validates a request into an object the API server will accept.
//
// Checked here and not left to admission, because the two refusals do not
// arrive in the same place. A field this function rejects is answered in the
// response to the request that set it; a field admission rejects is answered
// to Argo CD, minutes later, in a sync somebody has to go and read. The CRD
// keeps its rules either way — this is not a second copy of them, it is the
// subset that can be checked before a commit exists.
func databaseFrom(
	name, engine, image, storage, database, username string,
	wantBackups bool, backups func() (string, string, int32, bool),
) (platformv1alpha1.Database, error) {
	var zero platformv1alpha1.Database
	switch {
	case name == "":
		return zero, errors.New("a database needs a name")
	case !databaseName.MatchString(name):
		return zero, fmt.Errorf(
			"%q is not a usable name: lower-case letters, digits and dashes, "+
				"starting and ending with a letter or digit", name)
	case len(name) > 63:
		return zero, fmt.Errorf("%q is longer than the 63 characters a name may have", name)
	}

	kind := platformv1alpha1.DatabaseEngine(engine)
	if engine == "" {
		kind = platformv1alpha1.EnginePostgres
	}
	if kind != platformv1alpha1.EnginePostgres && kind != platformv1alpha1.EngineRedis {
		return zero, fmt.Errorf(
			"%q is not an engine this platform runs: postgres and redis are", engine)
	}

	if image == "" {
		return zero, errors.New("a database needs an image, pinned to a tag or a digest")
	}
	// The CRD's own rule, said in the response instead of in a sync. A tag that
	// moves is the thing the Database type spends its longest comment on:
	// PostgreSQL will not start on a directory a newer major wrote, and Redis
	// destroys the way back from an upgrade rather than the upgrade.
	if strings.HasSuffix(image, ":latest") {
		return zero, errors.New(
			"the image must not use :latest — a tag that moves turns the next pod " +
				"restart into an outage nothing connects to the change that caused it")
	}
	if !strings.Contains(image, "@") {
		last := image[strings.LastIndex(image, "/")+1:]
		if !strings.Contains(last, ":") {
			return zero, fmt.Errorf(
				"the image %q carries no tag or digest, so what it runs would change "+
					"underneath the data directory", image)
		}
	}

	size, err := resource.ParseQuantity(storage)
	if err != nil {
		return zero, fmt.Errorf(
			"%q is not a size: 1Gi and 512Mi are. Storage has no default because it "+
				"cannot be made smaller later", storage)
	}
	if size.IsZero() || size.Sign() < 0 {
		return zero, errors.New("storage has to be greater than zero")
	}

	db := platformv1alpha1.Database{
		ObjectMeta: metav1.ObjectMeta{Name: name},
		Spec: platformv1alpha1.DatabaseSpec{
			Engine: kind, Image: image, Storage: size,
		},
	}
	if kind == platformv1alpha1.EnginePostgres {
		db.Spec.Database = database
		db.Spec.Username = username
	} else if database != "" || username != "" {
		// Refused rather than dropped. Redis numbers its databases and
		// authenticates without a user, so these two would be accepted,
		// committed, and mean nothing — a setting that is stored and does
		// nothing, which is the failure the settings page exists to prevent.
		return zero, errors.New(
			"redis takes no database name and no username: it numbers its databases " +
				"and authenticates with a password alone")
	}

	if wantBackups {
		if kind == platformv1alpha1.EngineRedis {
			return zero, errors.New(
				"redis has no backups here: the schedule would run pg_dump against a " +
					"server that has never heard of it")
		}
		schedule, backupSize, retain, rehearse := backups()
		archive, err := resource.ParseQuantity(backupSize)
		if err != nil || archive.IsZero() || archive.Sign() < 0 {
			return zero, fmt.Errorf(
				"the backup volume %q is not a size greater than zero; it is separate "+
					"from the data volume so that filling it cannot stop writes", backupSize)
		}
		db.Spec.Backup = &platformv1alpha1.DatabaseBackup{
			Schedule: schedule, RetainDays: retain, Storage: archive, Rehearse: &rehearse,
		}
	}
	return db, nil
}

// databaseName is the RFC 1123 label rule, which is what the API server will
// apply to the object this renders. Checked here so the refusal names the
// field rather than arriving from a sync three minutes later.
var databaseName = regexp.MustCompile(`^[a-z0-9]([-a-z0-9]*[a-z0-9])?$`)

// wireDatabase is what this endpoint says about a database.
//
// No password, no connection string, and no field that could carry one. The
// operator mints the credentials into a Secret the tenant's pods read; this
// endpoint reads committed manifests, which never held them. Saying so here
// rather than only in the panel, because the shape is the guarantee: there is
// no field to fill in wrongly later.
type wireDatabase struct {
	Name     string `json:"name"`
	Engine   string `json:"engine"`
	Image    string `json:"image"`
	Storage  string `json:"storage"`
	Database string `json:"database,omitempty"`
	Username string `json:"username,omitempty"`

	// Backups is absent rather than false-y when there are none, so that "no
	// backups" and "backups whose schedule this build cannot read" are not the
	// same JSON.
	Backups *wireDatabaseBackup `json:"backups,omitempty"`

	// SecretName is where the credentials are, not what they are. It is the
	// same thing the settings page shows for a secret: git carries which
	// Secret to look in and never the value.
	SecretName string `json:"secretName"`

	// Claims is what this object takes out of the namespace's budget.
	Claims int `json:"claims"`
}

// Named for the database rather than for backups in general: server/backup.go
// already has a wireBackup, and that one is a Workload's backup status.
type wireDatabaseBackup struct {
	Schedule   string `json:"schedule"`
	RetainDays int32  `json:"retainDays"`
	Storage    string `json:"storage"`
	Rehearse   bool   `json:"rehearse"`
}

func toWireDatabase(db platformv1alpha1.Database) wireDatabase {
	out := wireDatabase{
		Name: db.Name, Engine: string(db.Spec.Engine), Image: db.Spec.Image,
		Storage: db.Spec.Storage.String(), Database: db.Spec.Database,
		Username: db.Spec.Username, SecretName: db.Name, Claims: claimsFor(db),
	}
	if db.Spec.Backup != nil {
		out.Backups = &wireDatabaseBackup{
			Schedule: db.Spec.Backup.Schedule, RetainDays: db.Spec.Backup.RetainDays,
			Storage: db.Spec.Backup.Storage.String(),
			// A nil Rehearse is the CRD's default, which is on. Reported as what
			// it will do rather than as the absence of an answer.
			Rehearse: db.Spec.Backup.Rehearse == nil || *db.Spec.Backup.Rehearse,
		}
	}
	return out
}

// claimsFor is the measured cost of one Database, and it is not always one.
func claimsFor(db platformv1alpha1.Database) int {
	if db.Spec.Backup != nil && db.Spec.Engine != platformv1alpha1.EngineRedis {
		return claimsForData + claimsForBackups
	}
	return claimsForData
}

// committedDatabases reads the databases already in the tenant's repository.
//
// From git and not from the cluster, and the reason is what a person does next.
// The cluster answers what Argo CD has applied so far; a database created a
// moment ago is not in it, so a create followed by a list would show nothing
// and read as a create that failed. Git answers what this app is configured to
// have, which is the question this endpoint's own writes change.
func committedDatabases(files map[string][]byte) []platformv1alpha1.Database {
	var out []platformv1alpha1.Database
	for _, name := range slices.Sorted(maps.Keys(files)) {
		if db, ok := manifest.ParseDatabase(files[name]); ok {
			out = append(out, db)
		}
	}
	return out
}

// listDatabases answers what this app environment is configured to run.
func listDatabases(g guard, st stores) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, ref, ok := g.admit(w, r, authz.ActionAppView)
		if !ok {
			return
		}
		place, files, done := readCommitted(w, r, st, ref)
		if !done {
			return
		}
		dbs := committedDatabases(files)
		wire := make([]wireDatabase, 0, len(dbs))
		used := 0
		for _, db := range dbs {
			wire = append(wire, toWireDatabase(db))
			used += claimsFor(db)
		}
		writeJSON(w, map[string]any{
			"databases":    wire,
			fieldNamespace: place.Namespace,
			// Stated, not enforced. See the comment on claimBudget for the
			// three reasons this is a fact about this directory rather than a
			// prediction about the namespace.
			"claims": map[string]any{
				"budget":            claimBudget,
				"usedByTheseHere":   used,
				"sharedWithTheApps": true,
			},
		})
	})
}

// readCommitted clones once and hands back the directory.
func readCommitted(
	w http.ResponseWriter, r *http.Request, st stores, ref evidence.Ref,
) (placement.Placement, map[string][]byte, bool) {
	place, err := st.placement.Get(r.Context(), ref.TenantID, ref.App, ref.Env)
	switch {
	case errors.Is(err, placement.ErrNotFound):
		problem(w, http.StatusNotFound, "this app and environment have no repository configured yet")
		return placement.Placement{}, nil, false
	case err != nil:
		problem(w, http.StatusInternalServerError, "reading the placement failed")
		return placement.Placement{}, nil, false
	}
	if st.writer == nil {
		problem(w, http.StatusNotImplemented, "this installation has no git writer configured")
		return placement.Placement{}, nil, false
	}
	method, err := st.gitAuth.For(place.RepoURL)
	if err != nil {
		problem(w, http.StatusInternalServerError, err.Error())
		return placement.Placement{}, nil, false
	}
	files, err := st.writer.Files(r.Context(), gitwrite.Target{
		RepoURL: place.RepoURL, Branch: place.Branch, Dir: place.Path, Auth: method,
	})
	if err != nil {
		problem(w, http.StatusBadGateway, "the repository could not be read: "+err.Error())
		return placement.Placement{}, nil, false
	}
	return place, files, true
}

// createDatabase commits a Database into the tenant's repository.
func createDatabase(g guard, st stores) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		sub, ref, ok := g.admit(w, r, authz.ActionDatabaseCreate)
		if !ok {
			return
		}
		var req struct {
			Name     string `json:"name"`
			Engine   string `json:"engine"`
			Image    string `json:"image"`
			Storage  string `json:"storage"`
			Database string `json:"database"`
			Username string `json:"username"`
			Backups  *struct {
				Schedule   string `json:"schedule"`
				RetainDays int32  `json:"retainDays"`
				Storage    string `json:"storage"`
				Rehearse   bool   `json:"rehearse"`
			} `json:"backups"`
		}
		if err := json.NewDecoder(
			http.MaxBytesReader(w, r.Body, maxRequestBody)).Decode(&req); err != nil {
			problem(w, http.StatusBadRequest, "reading the request: "+err.Error())
			return
		}

		db, err := databaseFrom(req.Name, req.Engine, req.Image, req.Storage,
			req.Database, req.Username, req.Backups != nil, func() (string, string, int32, bool) {
				if req.Backups == nil {
					return "", "", 0, false
				}
				return req.Backups.Schedule, req.Backups.Storage, req.Backups.RetainDays, req.Backups.Rehearse
			})
		if err != nil {
			problem(w, http.StatusBadRequest, err.Error())
			return
		}

		commitChange(w, r, st, sub, ref, "add database "+db.Name,
			func(place placement.Placement) renderFunc {
				return func(_ string, current map[string][]byte) (map[string][]byte, error) {
					for _, existing := range committedDatabases(current) {
						if existing.Name == db.Name {
							return nil, fmt.Errorf(
								"%w: this environment already has a database called %q",
								errDatabaseExists, db.Name)
						}
					}
					db.Namespace = place.Namespace
					body, err := manifest.RenderDatabase(db)
					if err != nil {
						return nil, err
					}
					// Everything already committed, plus one file. A render
					// that returned only the new database would have gitwrite
					// remove the workload beside it — Owns says these files are
					// ours, and ours is exactly what it is willing to delete.
					out := make(map[string][]byte, len(current)+1)
					maps.Copy(out, current)
					out[manifest.FileFor("Database", db.Name)] = body
					return out, nil
				}
			},
			// No removal: this render returns everything already committed and
			// one file more, so there is nothing for gitwrite to withdraw, and
			// a caller that does not need the power should not hold it.
			commitOptions{})
	})
}

// deleteDatabase withdraws the manifest, and says what that does to the data.
//
// "Removed" and "the data is gone" are deliberately not the same sentence,
// because they are not the same event and the difference is measurable:
//
//   - The Database object goes, and with it the StatefulSet, the Service and
//     the Secret holding the credentials. Those are owned, so the garbage
//     collector takes them.
//   - Both volumes stay. Kubernetes leaves a StatefulSet's volumeClaimTemplate
//     claims behind on purpose, and the operator creates the archive claim
//     unowned for exactly this reason — its comment says an owner reference
//     "would have the garbage collector delete the archives when the Database
//     is deleted".
//   - So the claims stay spent. Removing a database does not give the
//     namespace its budget back, which is the opposite of what somebody
//     deleting one to make room would assume.
//   - And a database recreated under the same name attaches to the volumes
//     that are still there. That is a feature where it is wanted — the
//     archives survive — and a surprise where it is not: the new database is
//     not empty.
//
// None of that is this endpoint's to change. What it can do is say it, in the
// response, at the moment somebody asks for it.
func deleteDatabase(g guard, st stores) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		sub, ref, ok := g.admit(w, r, authz.ActionDatabaseDelete)
		if !ok {
			return
		}
		name := r.PathValue("database")
		commitChange(w, r, st, sub, ref, "remove database "+name,
			func(_ placement.Placement) renderFunc {
				return func(_ string, current map[string][]byte) (map[string][]byte, error) {
					file := manifest.FileFor("Database", name)
					if _, ok := current[file]; !ok {
						return nil, fmt.Errorf("%w: %q", errNoSuchDatabase, name)
					}
					out := make(map[string][]byte, len(current))
					maps.Copy(out, current)
					delete(out, file)
					if len(out) == 0 {
						// gitwrite refuses an empty render, and it is right to:
						// an app directory with no files is indistinguishable
						// from a render that went wrong. Reported as itself.
						return nil, errLastManifest
					}
					return out, nil
				}
			},
			commitOptions{Note: dataAfterDelete, MayRemove: true})
	})
}

// dataAfterDelete is what the response says, and every clause in it was
// measured in the operator rather than assumed.
const dataAfterDelete = "the database is withdrawn: its StatefulSet, Service and " +
	"credentials Secret are owned by it and go with it. The data volume and the backup " +
	"volume do not — Kubernetes leaves a StatefulSet's claims behind on purpose, and the " +
	"archive claim is created unowned so the garbage collector cannot take it. So the " +
	"data is still here, it still counts against the namespace's four claims, and a " +
	"database created again under this name will attach to it rather than start empty."
